### PR TITLE
CPython engine marshals python classes to lists when the expectation is to get a handle.

### PR DESCRIPTION
**Pull request by mjkkirschner**
### Purpose

While trying to get data-shapes working with the cpython engine it was discovered that some of the data-shapes python classes implement `__iter__/__getitem__` and sequence protocols.

These classes are marshaled into dynamo as lists, but for use in data-shapes they need to be marshaled as handles only.

To enable this workflow I've added a new "builtin" 👋   attribute the class author can choose to add to their class. `__dynamoskipconversion__`. If this attribute exists at all (the return value does not matter) then the object will skip all conversion attempts and immediately be marshaled as a handle.

For dataShapes this means a one line code change to get this class working.

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [x] This PR modifies some build requirements and the readme is updated
- [ ] **When/If this PR is approved I will update the python3 wiki with information on this attribute.**

### CODE REVIEW

File: test/Libraries/DynamoPythonTests/PythonEvalTests.cs [hunk 1]
```diff
@@ -212,6 +212,46 @@ class myobj:
...
+        return iter([0,1,2,3])
+    def __getitem__(self,key):
+        return key
+
+o = notiterable()
+OUT = o
+";
+            var empty = new ArrayList();
+           
+           Assert.IsInstanceOf(typeof(IList),DSCPython.CPythonEvaluator.EvaluatePythonScript(code, empty, empty));
```
Comment (mmisol) [inline test/Libraries/DynamoPythonTests/PythonEvalTests.cs:249]: Maybe also assert the value for this case?

Comment (mjkkirschner) [reply test/Libraries/DynamoPythonTests/PythonEvalTests.cs:249]: will do

File: test/Libraries/DynamoPythonTests/PythonEvalTests.cs [hunk 2]
```diff
@@ -212,6 +212,46 @@ class myobj:
             });
         }
 
+        [Test]
+        public void PythonObjectWithDynamoSKIPisNotMarshaled()
```
[the remark this answers is not in the document]
Comment (mjkkirschner) [reply test/Libraries/DynamoPythonTests/PythonEvalTests.cs:216]: no good reason, will change it.

File: src/Libraries/DSCPython/CPythonEvaluator.cs [hunk 3]
```diff
@@ -459,6 +467,11 @@ public static DataMarshaler OutputMarshaler
             }
         }
 
+        private static bool IsMarkedToSkipConversion(PyObject pyObj)
+        {
+            return pyObj.HasAttr("__dynamoskipconversion__");
```
Comment (aparajit-pratap) [inline src/Libraries/DSCPython/CPythonEvaluator.cs:472]: Can we make this string a constant member variable of the class?

File: test/Libraries/DynamoPythonTests/PythonEvalTests.cs [hunk 4]
```diff
@@ -212,6 +214,47 @@ class myobj:
...
+        return iter([0,1,2,3])
+    def __getitem__(self,key):
+        return key
+
+o = notiterable()
+OUT = o
+";
+            var empty = new ArrayList();
+
+            Assert.IsInstanceOf(typeof(IList), DSCPython.CPythonEvaluator.EvaluatePythonScript(code, empty, empty));
```
Comment (aparajit-pratap) [inline test/Libraries/DynamoPythonTests/PythonEvalTests.cs:251]: Why not just store the result of `DSCPython.CPythonEvaluator.EvaluatePythonScript(code, empty, empty)` in a local var and use it in both places?

File: src/Libraries/DSCPython/CPythonEvaluator.cs [hunk 5]
```diff
@@ -387,6 +387,14 @@ public static DataMarshaler OutputMarshaler
                             {
                                 return outputMarshaler.Marshal(clrObj);
                             }
+
+                            if (IsMarkedToSkipConversion(pyObj))
+                            {
+                                var globalScope = PyScopeManager.Global.Get(globalScopeName);
+                                globalScope.Set(pyObj.Handle.ToString(), pyObj);
+                                return new DynamoCPythonHandle(pyObj.Handle);
```
Comment (aparajit-pratap) [inline src/Libraries/DSCPython/CPythonEvaluator.cs:395]: Minor but these 3 lines of code are repeated in 2 places - here and further down below. Might be worth refactoring into a helper method?

File: test/Libraries/DynamoPythonTests/PythonEvalTests.cs (hunk 4 above)
Comment (mjkkirschner) [reply test/Libraries/DynamoPythonTests/PythonEvalTests.cs:251]: sure, can do.

File: src/Libraries/DSCPython/CPythonEvaluator.cs (hunk 5 above)
Comment (mjkkirschner) [reply src/Libraries/DSCPython/CPythonEvaluator.cs:395]: wil fix.

**Pull request merged**